src/index.js

```diff
@@ -1 +1,34 @@
-export default () => true
+import defaultsDeep from 'lodash.defaultsdeep';
+import pickBy from 'lodash.pickby';
+import makeOkCheck from './make-ok-check';
+import buildOptions from './build-options';
+import resolveUrl from './resolve-url';
+
+class Stipulate {
+
+  constructor(options) {
+    this.baseOptions = options;
+  }
+
+  beforeRequest(url, options) {
+    return [url, options];
+  }
+
+  afterResponse(response) {
+    return response;
+  }
+
+  send(urlString, options, query) {
+    const config = buildOptions(options, this.baseOptions);
+    const url = resolveUrl(urlString, query);
+
+    const requestArguments = this.beforeRequest(url, config);
+    const checkOk = makeOkCheck(config);
+
+    return fetch(...requestArguments)
+      .then(checkOk)
+      .then(this.afterResponse);
```

Comment:
> the problem with your examples is they are function calls that, presumably, have some sort of side effect that is indirectly important is some way. (i.e. you're not using the return value of doSomethingElseBeforeDoSomething())

I think you missed the point. The functions don't have to have side effects, they can just as easily manipulate the value.

> that's achievable, and more flexible, with the current implementation and using super, because whatever changes i want to make can be in the before and after stub functions, and users that extend to override those functions can choose to super in my additions or not, at their discretion.

Then these aren't hooks. They're class methods intended to be overwritten. Hooks don't necessitate a class extension, and while you can say that people can manipulate the methods directly to do what they want, you can't do that AND get the benefit of changes to stipulate. Right now, there is no reason to use `super` because the functions are literally placeholders. If they're suddenly not placeholders, then users that didn't use `super` are SOL and can't get the benefit of whatever changes you made.

To me, that seems like poor design.



> I don't see the strength you're referring to. I see extraneous type-checking and masking of functions to keep users from taking advantage of class inheritance, which is part of es2015.

There are issues with class inheritance from a design standpoint; not every user is going to want that. I don't see the point in shoehorning them into class inheritance or mucking with the base class to get the functionality you want to give them.